vcd/resource_vcd_org_vdc_access_control.go

```diff
@@ -0,0 +1,228 @@
+	if err != nil {
+		return diag.Errorf("error when setting VDC control access parameters - %s", err)
+	}
+
+	d.SetId(vdc.Vdc.ID)
+	return resourceVcdVdcAccessControlRead(ctx, d, meta)
+}
+
+func resourceVcdVdcAccessControlRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	org, err := vcdClient.GetOrgFromResource(d)
+	if err != nil {
+		return diag.Errorf("error while reading Org - %s", err)
+	}
+
+	vdc, err := org.GetVDCById(d.Id(), false)
+	if err != nil {
+		if govcd.IsNotFound(err) {
+			d.SetId("")
+		} else {
+			return diag.Errorf("error while reading VDC - %s", err)
+		}
+	}
+
+	controlAccessParams, err := vdc.GetControlAccess(true)
+	if err != nil {
+		return diag.Errorf("error getting control access parameters - %s", err)
+	}
+
+	dSet(d, "shared_with_everyone", controlAccessParams.IsSharedToEveryone)
+	if controlAccessParams.EveryoneAccessLevel != nil {
+		dSet(d, "everyone_access_level", *controlAccessParams.EveryoneAccessLevel)
+	} else {
+		dSet(d, "everyone_access_level", "")
+	}
+
+	if controlAccessParams.AccessSettings != nil {
+		accessControlListSet, err := accessControlListToSharedSet(controlAccessParams.AccessSettings.AccessSetting)
+		if err != nil {
+			return diag.Errorf("error converting slice AccessSetting into set - %s", err)
+		}
+
+		err = d.Set("shared_with", accessControlListSet)
+		if err != nil {
+			return diag.Errorf("error setting shared_with attribute - %s", err)
+		}
+	}
+
+	return nil
+}
+
+func resourceVcdVdcAccessControlDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	// When deleting VDC access control, VDC won't be share with anyone, neither everyone not any single user/group
+	vcdClient := meta.(*VCDClient)
+
+	org, _, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		diag.Errorf("error when retrieving Org - %s", err)
+	}
+
+	vdc, err := org.GetVDCById(d.Id(), false)
+	if err != nil {
+		diag.Errorf("error when retrieving VDC - %s", err)
+	}
+
+	_, err = vdc.DeleteControlAccess(true)
+	if err != nil {
+		return diag.Errorf("error when deleting VDC access control - %s", err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func resourceVcdVdcAccessControlImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	resourceURI := strings.Split(d.Id(), ImportSeparator)
+	if len(resourceURI) != 2 {
+		return nil, fmt.Errorf("resource name must be specified as org.vdc")
+	}
+
+	orgName, vdcName := resourceURI[0], resourceURI[1]
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdc(orgName, vdcName)
+	if err != nil {
+		return nil, fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	dSet(d, "org", orgName)
+	dSet(d, "vdc", vdcName)
+	d.SetId(vdc.Vdc.ID)
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func checkParamsVdcAccessControl(isSharedWithEveryone bool, everyoneAccessLevelSet bool, sharedList []interface{}) error {
```

Comment:
Have a look at schema validation attributes `ConflictsWith`, `ExactlyOneOf`, `AtLeastOneOf` `RequiredWith` - they might save you these manual validations and also - if you use these schema attribute fields - validation occurs natively in Terraform (will be checked using `terraform plan`)